File: examples/julia/basic_denoising.jl
#=
Promethium.jl Examples
Basic Denoising Example

This script demonstrates seismic data denoising using Promethium.jl.
It shows loading data, applying filters, and evaluating reconstruction quality.

Usage:
    julia basic_denoising.jl
=#

using Promethium
using Random
using Statistics
using Printf

println("Promethium.jl - Basic Denoising Example")
println("=" ^ 50)

# -----------------------------------------------------------------------------
# 1. Generate Synthetic Data
# -----------------------------------------------------------------------------

println("\n1. Generating synthetic seismic data...")

# Set random seed for reproducibility
Random.seed!(42)

n_traces = 100
n_samples = 500
sample_rate = 0.004  # 4 ms

# Generate time axis
time_axis = collect(0:sample_rate:(n_samples-1)*sample_rate)

# Create clean synthetic data with reflections
clean_data = zeros(Float64, n_traces, n_samples)

# Ricker wavelet function
function ricker_wavelet(t, f0)
    return (1 - 2 * (π * f0 * t)^2) * exp(-(π * f0 * t)^2)
end

# Add synthetic reflections
reflection_times = [0.2, 0.5, 0.8, 1.2]
f0 = 30.0  # Dominant frequency

for i in 1:n_traces
    for rt in reflection_times
        idx = argmin(abs.(time_axis .- rt))
        amp = 0.5 + 0.5 * rand()
        
        # Add Ricker wavelet
        wavelet_half = 50
        for j in max(1, idx-wavelet_half):min(n_samples, idx+wavelet_half)
            t = (j - idx) * sample_rate
            clean_data[i, j] += amp * ricker_wavelet(t, f0)
        end
    end
end

# Add noise
noise_level = 0.3
noisy_data = clean_data .+ noise_level .* randn(n_traces, n_samples)

@printf("  Traces: %d\n", n_traces)
@printf("  Samples: %d\n", n_samples)
@printf("  Sample rate: %.1f ms\n", sample_rate * 1000)
@printf("  Noise level: %.2f\n", noise_level)

# -----------------------------------------------------------------------------
# 2. Create SeismicDataset
# -----------------------------------------------------------------------------

println("\n2. Creating SeismicDataset objects...")

# Create dataset from noisy data
noisy_dataset = SeismicDataset(
    noisy_data,
    sample_rate=sample_rate,
    metadata=Dict(
        "source" => "synthetic",
        "description" => "Noisy synthetic shot gather"
    )
)

println(noisy_dataset)

# -----------------------------------------------------------------------------
# 3. Apply Wiener Filter
# -----------------------------------------------------------------------------

println("\n3. Applying Wiener filter denoising...")

# Apply Wiener filter
wiener_result = wiener_filter(noisy_dataset, noise_power=noise_level^2)

println("   Wiener filter applied successfully.")

# -----------------------------------------------------------------------------
# 4. Apply Matrix Completion (if traces are missing)
# -----------------------------------------------------------------------------

println("\n4. Demonstrating matrix completion for missing traces...")

# Create data with missing traces
missing_ratio = 0.3
n_missing = round(Int, n_traces * missing_ratio)
missing_indices = sort(randperm(n_traces)[1:n_missing])

missing_data = copy(noisy_data)
missing_data[missing_indices, :] .= 0.0

# Create mask
mask = ones(Bool, n_traces, n_samples)
mask[missing_indices, :] .= false

# Apply matrix completion
@printf("   Removing %d traces (%.0f%% missing)...\n", n_missing, missing_ratio * 100)

# Note: matrix_completion_ista is from the Promethium package
recovered_data = matrix_completion_ista(
    missing_data,
    mask,
    lambda=0.1,
    max_iter=100
)

println("   Matrix completion applied successfully.")

# -----------------------------------------------------------------------------
# 5. Compute Quality Metrics
# -----------------------------------------------------------------------------

println("\n" * "-" ^ 50)
println("Quality Metrics")
println("-" ^ 50)

# SNR function
function compute_snr(reference, estimate)
    signal_power = sum(reference.^2)
    noise_power = sum((reference .- estimate).^2)
    return 10 * log10(signal_power / noise_power)
end

# MSE function
function compute_mse(reference, estimate)
    return mean((reference .- estimate).^2)
end

# Compute metrics for Wiener filter
wiener_data = wiener_result.traces
snr_noisy = compute_snr(clean_data, noisy_data)
snr_wiener = compute_snr(clean_data, wiener_data)

@printf("Original noisy SNR:    %.2f dB\n", snr_noisy)
@printf("After Wiener filter:   %.2f dB\n", snr_wiener)
@printf("  Improvement:         %.2f dB\n", snr_wiener - snr_noisy)

# Compute metrics for matrix completion (only on recovered traces)
if !isempty(missing_indices)
    snr_missing = compute_snr(clean_data[missing_indices, :], recovered_data[missing_indices, :])
    @printf("\nRecovered traces SNR:  %.2f dB\n", snr_missing)
end

# MSE
mse_noisy = compute_mse(clean_data, noisy_data)
mse_wiener = compute_mse(clean_data, wiener_data)

@printf("\nMSE (noisy):          %.4e\n", mse_noisy)
@printf("MSE (Wiener):         %.4e\n", mse_wiener)

# -----------------------------------------------------------------------------
# 6. Save Results (optional)
# -----------------------------------------------------------------------------

println("\n" * "-" ^ 50)
println("Saving results...")

# Create output directory
output_dir = "results"
mkpath(output_dir)

# Save results as NPY (if NPZ.jl available) or JLD2
try
    using JLD2
    
    results_file = joinpath(output_dir, "denoising_results.jld2")
    @save results_file clean_data noisy_data wiener_data recovered_data
    println("Results saved to: $results_file")
catch e
    # Fallback: save as delimited text
    results_file = joinpath(output_dir, "denoising_results.csv")
    # Just save a summary
    println("JLD2 not available. Skipping file save.")
end

println("\nExample completed successfully.")

File: examples/scala/BasicDenoising.scala
/**
 * Promethium Scala Examples
 * Basic Denoising Example
 *
 * This example demonstrates seismic data denoising using promethium-scala.
 * It shows how to load data, apply denoising algorithms, and evaluate results.
 *
 * To run:
 *   sbt "runMain examples.BasicDenoising"
 */
package examples

import io.github.olaflaitinen.promethium._
import io.github.olaflaitinen.promethium.core.{SeismicDataset, VelocityModel}
import io.github.olaflaitinen.promethium.signal.{WienerFilter, BandpassFilter}
import io.github.olaflaitinen.promethium.metrics.{SNR, MSE, SSIM}
import io.github.olaflaitinen.promethium.pipelines.SeismicRecoveryPipeline

import scala.util.Random
import breeze.linalg._
import breeze.numerics._

object BasicDenoising {

  def main(args: Array[String]): Unit = {
    println("Promethium Scala - Basic Denoising Example")
    println("=" * 50)

    // -------------------------------------------------------------------------
    // 1. Generate Synthetic Data
    // -------------------------------------------------------------------------
    println("\n1. Generating synthetic seismic data...")

    val nTraces = 100
    val nSamples = 500
    val sampleRate = 0.004  // 4 ms
    val random = new Random(42)

    // Generate clean synthetic data
    val cleanData = DenseMatrix.zeros[Double](nTraces, nSamples)
    val reflectionTimes = Seq(0.2, 0.5, 0.8, 1.2)
    val f0 = 30.0  // Dominant frequency

    def rickerWavelet(t: Double, freq: Double): Double = {
      val piF = math.Pi * freq * t
      (1 - 2 * piF * piF) * math.exp(-piF * piF)
    }

    for (i <- 0 until nTraces) {
      for (rt <- reflectionTimes) {
        val idx = (rt / sampleRate).toInt
        val amp = 0.5 + 0.5 * random.nextDouble()

        for (j <- math.max(0, idx - 50) until math.min(nSamples, idx + 50)) {
          val t = (j - idx) * sampleRate
          cleanData(i, j) += amp * rickerWavelet(t, f0)
        }
      }
    }

    // Add noise
    val noiseLevel = 0.3
    val noisyData = cleanData + DenseMatrix.rand[Double](nTraces, nSamples) * noiseLevel

    println(s"  Traces: $nTraces")
    println(s"  Samples: $nSamples")
    println(s"  Sample rate: ${sampleRate * 1000} ms")
    println(s"  Noise level: $noiseLevel")

    // -------------------------------------------------------------------------
    // 2. Create SeismicDataset
    // -------------------------------------------------------------------------
    println("\n2. Creating SeismicDataset...")

    val noisyDataset = SeismicDataset(
      traces = noisyData,
      sampleRate = sampleRate,
      metadata = Map(
        "source" -> "synthetic",
        "description" -> "Noisy synthetic shot gather"
      )
    )

    println(noisyDataset)

    // -------------------------------------------------------------------------
    // 3. Apply Wiener Filter
    // -------------------------------------------------------------------------
    println("\n3. Applying Wiener filter denoising...")

    val wienerFilter = WienerFilter(noisePower = noiseLevel * noiseLevel)
    val denoisedDataset = wienerFilter.apply(noisyDataset)

    println("   Wiener filter applied successfully.")

    // -------------------------------------------------------------------------
    // 4. Apply Bandpass Filter
    // -------------------------------------------------------------------------
    println("\n4. Applying bandpass filter...")

    val bandpassFilter = BandpassFilter(
      lowFreq = 5.0,
      highFreq = 60.0,
      sampleRate = 1.0 / sampleRate
    )
    val bandpassResult = bandpassFilter.apply(noisyDataset)

    println("   Bandpass filter applied successfully.")

    // -------------------------------------------------------------------------
    // 5. Compute Quality Metrics
    // -------------------------------------------------------------------------
    println("\n" + "-" * 50)
    println("Quality Metrics")
    println("-" * 50)

    // SNR calculation
    def computeSNR(reference: DenseMatrix[Double], estimate: DenseMatrix[Double]): Double = {
      val signalPower = sum(reference *:* reference)
      val diff = reference - estimate
      val noisePower = sum(diff *:* diff)
      10 * math.log10(signalPower / noisePower)
    }

    // MSE calculation
    def computeMSE(reference: DenseMatrix[Double], estimate: DenseMatrix[Double]): Double = {
      val diff = reference - estimate
      sum(diff *:* diff) / (reference.rows * reference.cols)
    }

    val snrNoisy = computeSNR(cleanData, noisyData)
    val snrDenoised = computeSNR(cleanData, denoisedDataset.traces)

    println(f"Original noisy SNR:   $snrNoisy%.2f dB")
    println(f"After Wiener filter:  $snrDenoised%.2f dB")
    println(f"  Improvement:        ${snrDenoised - snrNoisy}%.2f dB")

    val mseNoisy = computeMSE(cleanData, noisyData)
    val mseDenoised = computeMSE(cleanData, denoisedDataset.traces)

    println(f"\nMSE (noisy):         $mseNoisy%.4e")
    println(f"MSE (denoised):      $mseDenoised%.4e")

    // -------------------------------------------------------------------------
    // 6. Run Pipeline
    // -------------------------------------------------------------------------
    println("\n" + "-" * 50)
    println("Running recovery pipeline...")

    val pipeline = SeismicRecoveryPipeline.fromPreset("wiener")
    val pipelineResult = pipeline.run(noisyDataset)

    println("Pipeline completed successfully.")

    // -------------------------------------------------------------------------
    // 7. Summary
    // -------------------------------------------------------------------------
    println("\n" + "=" * 50)
    println("Example completed successfully.")
    println(s"SNR improvement: ${snrDenoised - snrNoisy}%.2f dB")
  }
}
